numpy/core/src/multiarray/dtype_traversal.c

```diff
@@ -0,0 +1,455 @@
+                    return -1;
+                }
+            }
+            N -= blocksize;
+            data += blocksize * stride;
+        }
+        else {
+            for (i = 0; i < field_count; ++i) {
+                single_field_clear_data field = d->fields[i];
+                if (field.info.func(traverse_context,
+                        field.info.descr, data + field.src_offset,
+                        N, stride, field.info.auxdata) < 0) {
+                    return -1;
+                }
+            }
+            return 0;
+        }
+    }
+}
+
+
+static int
+get_clear_fields_transfer_function(
+        void *traverse_context, PyArray_Descr *dtype, int NPY_UNUSED(aligned),
+        npy_intp stride, traverse_loop_function **out_func,
+        NpyAuxData **out_auxdata, NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    PyObject *names, *key, *tup, *title;
+    PyArray_Descr *fld_dtype;
+    npy_int i, structsize;
+    Py_ssize_t field_count;
+
+    names = dtype->names;
+    field_count = PyTuple_GET_SIZE(dtype->names);
+
+    /* Over-allocating here: less fields may be used */
+    structsize = (sizeof(fields_clear_data) +
+                    field_count * sizeof(single_field_clear_data));
+    /* Allocate the data and populate it */
+    fields_clear_data *data = PyMem_Malloc(structsize);
+    if (data == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    data->base.free = &fields_clear_data_free;
+    data->base.clone = &fields_clear_data_clone;
+    data->field_count = 0;
+
+    single_field_clear_data *field = data->fields;
+    for (i = 0; i < field_count; ++i) {
+        int offset;
+
+        key = PyTuple_GET_ITEM(names, i);
+        tup = PyDict_GetItem(dtype->fields, key);
+        if (!PyArg_ParseTuple(tup, "Oi|O", &fld_dtype, &offset, &title)) {
+            NPY_AUXDATA_FREE((NpyAuxData *)data);
+            return -1;
+        }
+        if (PyDataType_REFCHK(fld_dtype)) {
+            NPY_ARRAYMETHOD_FLAGS clear_flags;
+            if (get_clear_function(
+                    traverse_context, fld_dtype, 0,
+                    stride, &field->info, &clear_flags) < 0) {
+                NPY_AUXDATA_FREE((NpyAuxData *)data);
+                return -1;
+            }
+            *flags = PyArrayMethod_COMBINED_FLAGS(*flags, clear_flags);
+            field->src_offset = offset;
+            data->field_count++;
+            field++;
+        }
+    }
+
+    *out_func = &traverse_fields_function;
+    *out_auxdata = (NpyAuxData *)data;
+
+    return 0;
+}
+
+
+typedef struct {
+    NpyAuxData base;
+    npy_intp count;
+    NPY_traverse_info info;
+} subarray_clear_data;
+
+
+/* transfer data free function */
+static void
+subarray_clear_data_free(NpyAuxData *data)
+{
+    subarray_clear_data *d = (subarray_clear_data *)data;
+
+    NPY_traverse_info_xfree(&d->info);
+    PyMem_Free(d);
+}
+
+
+/* transfer data copy function */
+static NpyAuxData *
+subarray_clear_data_clone(NpyAuxData *data)
```

Comment:
This is never used?

Reply:
Yes, this one actually seems unreachable, added a copy.  I think I covered most of the others (but its complicated, will have to see what is actually covered).